bindings/src/test/java/io/pravega/storage/extendeds3/S3Mock.java

```diff
@@ -0,0 +1,315 @@
+                        object.setSize((long) e.getValue().content.getLength());
+                        object.setLastModified(new Date(System.currentTimeMillis()));
+                        return object;
+                    })
+                    .collect(Collectors.toList());
+        }
+        result.setObjects(list);
+        result.setMaxKeys(list.size());
+        result.setBucketName(bucketName);
+        result.setPrefix(prefix);
+        return result;
+    }
+
+    public S3ObjectMetadata getObjectMetadata(String bucketName, String key) {
+        String objectName = getObjectName(bucketName, key);
+        S3ObjectMetadata metadata = new S3ObjectMetadata();
+        synchronized (this.objects) {
+            ObjectData od = this.objects.get(objectName);
+            if (od == null) {
+                throw new S3Exception("NoSuchKey", 404, "NoSuchKey", "");
+            }
+            metadata.setContentLength((long) od.content.getLength());
+            metadata.setLastModified(new Date(System.currentTimeMillis()));
+        }
+        return metadata;
+    }
+
+    public InputStream readObjectStream(String bucketName, String key, Range range) {
+        String objectName = getObjectName(bucketName, key);
+        synchronized (this.objects) {
+            ObjectData od = this.objects.get(objectName);
+            if (od == null) {
+                throw new S3Exception("NoSuchKey", 404, "NoSuchKey", "");
+            }
+
+            final int offset = (int) (long) range.getFirst();
+            final int length = (int) (range.getLast() - range.getFirst() + 1);
+            try {
+                return od.content.slice(offset, length).getReader();
+            } catch (Exception ex) {
+                throw new S3Exception("InvalidRange", org.apache.http.HttpStatus.SC_REQUESTED_RANGE_NOT_SATISFIABLE, "InvalidRange", "");
+            }
+        }
+    }
+
+    public String initiateMultipartUpload(String bucketName, String key) {
+        String objectName = getObjectName(bucketName, key);
+        synchronized (this.objects) {
+            this.multipartUploads.put(objectName, new HashMap<>());
+            return Integer.toString(this.multipartUploads.size());
+        }
+    }
+
+    public CompleteMultipartUploadResult completeMultipartUpload(CompleteMultipartUploadRequest request) {
+        String objectName = getObjectName(request.getBucketName(), request.getKey());
+        synchronized (this.objects) {
+            Map<Integer, CopyPartRequest> partMap = this.multipartUploads.get(objectName);
+            if (partMap == null) {
+                throw new S3Exception("NoSuchKey", 404, "NoSuchKey", "");
+            }
+
+            if (!this.objects.containsKey(objectName)) {
+                throw new S3Exception("NoSuchKey", 404, "NoSuchKey", "");
+            }
+
+            val partObjects = partMap.entrySet().stream().sorted(Comparator.comparingInt(Map.Entry::getKey))
```

Comment:
I am not sure what is the expected behavior here.
What if we have part 1, 3, 7 only.
Is it still a valid MPU request ? Or it is expected to throw in such cases?

Reply:
Fixed.